drivers: sensor: APDS9306: Lux conversion

Kampi opened the pull request:
- Add lux conversion to APDS-9306 driver
- Change settings of `gain`, `resolution` and `frequency` to index-based settings
- Add Device Tree overlay sample for APDS-9306
- Fix wrong board name in `light_polling` README
- Add value checks for the attribute set API call
- Remove the reading of the sensor attributes from the sensor and use buffered values instead
- Rename `frequency` property to `measurement period`

Closes #91104

Kampi commented on samples/sensor/apds9306/src/main.c:
@kartben I'm afraid I'm not following you. Can you please be more specific?

Kampi replied:
Makes sense :)
I will change the PR later

msalau commented on drivers/sensor/apds9306/apds9306.c:
```diff
@@ -128,6 +136,13 @@ static void apds9306_worker(struct k_work *p_work)
+	 * https://github.com/esphome/esphome/blob/dev/esphome/components/apds9306/apds9306.cpp#L144
+	 */
+	data->light = (data->light / avago_apds9306_gain[data->gain]) *
+		      (100.0f / avago_apds9306_frequency[data->frequency]);
```
There is another conversion formula which seems more logical to me:
https://github.com/gmarti/AsyncAPDS9306/blob/master/AsyncAPDS9306.cpp#L39
```c
  return ((float)raw / againValue) * (100.0 / atimeValue);
```
where `atimeValue` is not `ALS Measurement Rate` but `ALS Resolution/Bit Width` which is mentioned also as `ADC measurement time` and is specified both in bits and milliseconds.

And one more:
https://github.com/torvalds/linux/blob/master/drivers/iio/light/apds9306.c#L88

```c
/*
 * As per the datasheet, at HW Gain = 3x, Integration time 100mS (32x),
 * typical 2000 ADC counts are observed for 49.8 uW per sq cm (340.134 lux)
 * for apds9306 and 43 uW per sq cm (293.69 lux) for apds9306-065.
 * Assuming lux per count is linear across all integration time ranges.
 *
 * Lux = (raw + offset) * scale; offset can be any value by userspace.
 * HG = Hardware Gain; ITG = Gain by changing integration time.
 * Scale table by IIO GTS Helpers = (1 / HG) * (1 / ITG) * Multiplier.
 *
 * The Lux values provided in the datasheet are at ITG=32x and HG=3x,
 * at typical 2000 count for both variants of the device.
 *
 * Lux per ADC count at 3x and 32x for apds9306 = 340.134 / 2000
 * Lux per ADC count at 3x and 32x for apds9306-065 = 293.69 / 2000
 *
 * The Multiplier for the scale table provided to userspace:
 * IIO GTS scale Multiplier for apds9306 = (340.134 / 2000) * 32 * 3 = 16.326432
 * and for apds9306-065 = (293.69 / 2000) * 32 * 3 = 14.09712
 */
```

Kampi commented:
> There is a disconnect between `gain` property set in the devicetree and the `ALS_GAIN` register. Property defines the desired gain value, but register expects an index, not the value itself. The same issue exists for the `ALS_MEAS_RATE` register values (resolution and rate).
> 
> A possible solution:
> 
> 1. Reorder gain values in the yaml file to match the datasheet:
> 
> ```yaml
>   gain:
>     type: int
>     default: 1
>     enum:
>       - 1
>       - 3
>       - 6
>       - 9
>       - 18
> ```
> 
> 2. Save the gain enum's index in `config`:
> 
> ```c
> .gain_idx = DT_INST_ENUM_IDX(inst, gain), \
> ```
> 
> 3. Write index into the register:
> 
> ```c
> 	value = config->gain_idx;
> 	LOG_DBG("Write configuration 0x%x to register 0x%x", value, APDS9306_REGISTER_ALS_GAIN);
> 	if (i2c_reg_write_byte_dt(&config->i2c, APDS9306_REGISTER_ALS_GAIN, value)) {
> 		return -EFAULT;
> 	}
> ```
> 
> This should fix initialization, I believe.
> 
> I can't comment about the conversion part though.

Are you sure? I don´t work with the indexes. Instead I write the "real" values into the configuration

> There is a disconnect between `gain` property set in the devicetree and the `ALS_GAIN` register. Property defines the desired gain value, but register expects an index, not the value itself. The same issue exists for the `ALS_MEAS_RATE` register values (resolution and rate).
> 
> A possible solution:
> 
> 1. Reorder gain values in the yaml file to match the datasheet:
> 
> ```yaml
>   gain:
>     type: int
>     default: 1
>     enum:
>       - 1
>       - 3
>       - 6
>       - 9
>       - 18
> ```
> 
> 2. Save the gain enum's index in `config`:
> 
> ```c
> .gain_idx = DT_INST_ENUM_IDX(inst, gain), \
> ```
> 
> 3. Write index into the register:
> 
> ```c
> 	value = config->gain_idx;
> 	LOG_DBG("Write configuration 0x%x to register 0x%x", value, APDS9306_REGISTER_ALS_GAIN);
> 	if (i2c_reg_write_byte_dt(&config->i2c, APDS9306_REGISTER_ALS_GAIN, value)) {
> 		return -EFAULT;
> 	}
> ```
> 
> This should fix initialization, I believe.
> 
> I can't comment about the conversion part though.

Thanks! Yes, I´ve messed it up because I´ve designed the code to work with the values but the values aren´t equal to the bitmasks so I have to work with the indexes instead.

kartben commented on drivers/sensor/apds9306/apds9306.c:
```diff
@@ -252,13 +282,14 @@ static int apds9306_channel_get(const struct device *dev, enum sensor_channel ch
+	switch (channel) {
+	case SENSOR_CHAN_LIGHT:
+		value->val1 = data->light;
+		value->val2 = 0;
```
you're missing a `break` here :)

Kampi commented on drivers/sensor/apds9306/apds9306.c:
```diff
@@ -228,9 +232,9 @@ static int apds9306_sample_fetch(const struct device *dev, enum sensor_channel c
-	LOG_DBG("Measurement resolution: %u", resolution);
+	resolution_idx = (buffer >> 4) & 0x07;
+	delay = avago_apds9306_measurement_time[resolution_idx];
+	LOG_DBG("Measurement resolution: %u", resolution_idx);
```
That´s true. I´ve missed it, thanks!

msalau commented on drivers/sensor/apds9306/apds9306.c:
```diff
@@ -221,16 +224,9 @@ static int apds9306_sample_fetch(const struct device *dev, enum sensor_channel c
-	delay = apds9306_get_time_for_resolution(resolution);
-	LOG_DBG("Measurement resolution: %u", resolution);
+	delay = avago_apds9306_measurement_time[data->resolution_idx];
+	LOG_DBG("Measurement resolution: %u", data->resolution_idx);
```
The line will be confusing. I says resolution, but its index (register values) will be printed.

msalau commented on drivers/sensor/apds9306/apds9306.c:
```diff
@@ -221,16 +224,9 @@ static int apds9306_sample_fetch(const struct device *dev, enum sensor_channel c
-	LOG_DBG("Measurement resolution: %u", resolution);
+	delay = avago_apds9306_measurement_time[data->resolution_idx];
+	LOG_DBG("Measurement resolution: %u", data->resolution_idx);
 	LOG_DBG("Wait for %u ms", delay);
```
Actual delay will be longer by 100ms.

msalau commented on dts/bindings/sensor/avago,apds9306.yaml:
```diff
@@ -42,13 +42,13 @@ properties:
+      - 200
+      - 500
+      - 1000
+      - 2000
```
The property sets not a frequency value but a measurement period in milliseconds.
I think the property may be renamed to e.g. `measurement-period` (like [vishay,vcnl36825t](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/sensor/vishay%2Cvcnl36825t.yaml#L26))

Kampi replied:
Renaming the field causes issues with the Twister tests
https://github.com/zephyrproject-rtos/zephyr/actions/runs/15822735532/job/44595853980?pr=91105

msalau reviewed (commented):
Hello @Kampi 

Thanks!
Only few minor notes from me.

Kampi committed:
drivers: sensor: APDS9306: Add lux conversion

- Add lux conversion to APDS-9306 driver
- Change settings of gain, resolution and
frequency to index-based settings
- Add Device Tree overlay sample for APDS-9306
- Fix wrong board name in light_polling README
- Add value checks for the attribute set API call
- Remove the reading of the sensor attributes
from the sensor and use buffered values instead
- Rename `frequency` property to `measurement period`

Closes #91104

Signed-off-by: Daniel Kampert <DanielKampert@kampis-elektroecke.de>

msalau commented on drivers/sensor/apds9306/apds9306.c:
```diff
@@ -128,7 +125,18 @@ static void apds9306_worker(struct k_work *p_work)
+	 * https://github.com/gmarti/AsyncAPDS9306/blob/master/AsyncAPDS9306.cpp#L39
+	 */
+	double gain = avago_apds9306_gain[data->gain_idx];
+	double integrationtime = avago_apds9306_measurement_period[data->resolution_idx];
```
The `integrationtime` variable name doesn't follow `snake_case` naming convention. Please rename.

Kampi commented on drivers/sensor/apds9306/apds9306.c:
```diff
@@ -45,42 +45,39 @@
+
 struct apds9306_data {
 	uint32_t light;
+	uint8_t measurement_period_idx; /**< This field holds the index of the current */
+					/**< period measurement */
+	uint8_t gain_idx;       /**< This field holds the index of the current sampling gain. */
+	uint8_t resolution_idx; /**< This field holds the index of the current sampling */
+				/**<  resolution.*/
```
What do you mean? The CI/CD check doesn´t allow comments with `//`

msalau replied:
Is there a reason to use /**< */ instead of regular /* */ comments?

Kampi commented on drivers/sensor/apds9306/apds9306.c:
```diff
@@ -139,20 +147,30 @@ static int apds9306_attr_set(const struct device *dev, enum sensor_channel chann
 		return -ENOTSUP;
 	}
 
 	if (attribute == SENSOR_ATTR_SAMPLING_FREQUENCY) {
+		if (value->val1 > (AVAGO_APDS_9306_INTEGRATION_TIME_ARRAY_LENGTH - 1)) {
+			return -EINVAL;
+		}
```
Yeah but this value doesn´t exist so we need another constant here which means we split them up which makes sense here because we change the "purpose" of this value.

Kampi commented on drivers/sensor/apds9306/apds9306.c:
```diff
@@ -128,7 +126,18 @@ static void apds9306_worker(struct k_work *p_work)
 	data->light = sys_get_le24(buffer);
+	LOG_DBG("Last raw measurement: %u", data->light);
+
+	/* Based on
+	 * https://github.com/gmarti/AsyncAPDS9306/blob/master/AsyncAPDS9306.cpp#L39
+	 */
```
Hi @ubieda,
the same formula is used in the [Linux Kernel](https://github.com/tweaklogic/apds9306_backport/blob/9abab8238cdd2db4b06a7576f286481768589504/drivers/iio/light/apds9306.c#L112) and it´s a basic Gain to Timescale formula with the values derived from the sensor datasheet, as you can get it from [the APDS9309 datasheet](https://docs.broadcom.com/doc/AV02-3689EN) on page 4. So the only "secret" (if you can call it secret) is the factor which is set to `100` here. In the case of this sensor, `100` is chosen because of the reference measurement with 100 ms, as described in the [APDS-9306 datasheet](https://www.mouser.de/datasheet/2/678/AVGO_S_A0002854364_1-2574547.pdf) on page 4.

kartben replied:
@Kampi please simply reference information from the datasheet then ("see APDS-9306 datasheet, page 4", or whatever), not GPL code as this is asking for trouble.

Kampi replied:
Alright - I will expand the documentation and add the required references. I will also think about this `100`. It looks a bit suspicious because it doesn´t fit in the formula right now...

Kampi replied:
@kartben I´ve added a more understandable value, based on the values from the datasheet. So there is no violation anymore.

MaureenHelm commented:
> please wait a bit before accepting this PR. I want to check the results against a lux meter and I will do this this week.

Adding the `DNM` label to prevent merge. Please remove the label when you're done testing
